### PR TITLE
Self approve transfer

### PR DESCRIPTION
**Pull request by 0xdavinchee**
- interface added for selfApproveFor and selfTransferFrom
- functions added to CustomSuperTokenMock for testing
- tests added for both revert and successful cases for selfTransferFrom and selfApproveFor

### CODE REVIEW

**Review by hellwolf — COMMENTED** [review]
looks good.

@d10r please have a look too

File: packages/ethereum-contracts/contracts/mocks/CustomSuperTokenMock.sol [hunk 1]
```diff
@@ -63,4 +76,23 @@ contract CustomSuperTokenProxyMock is CustomSuperTokenBaseMock {
         // this makes msg.sender to self
         ISuperToken(address(this)).selfBurn(to, amount, userData);
     }
+    
+    // this function self calls transferFrom
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/mocks/CustomSuperTokenMock.sol:80]: perhaps from grammatical reading perspective, holder should be the first parameter for "transfer from"

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol [hunk 2]
```diff
@@ -315,6 +315,34 @@ interface ISuperToken is ISuperfluidToken, TokenInfo, IERC20, IERC777 {
...
 
+   /**
+    * @dev Transfer tokens from the `holder` to `recipient`. If `spender`
+    * isn't the same as `holder`, checks if `holder` has allowance to 
+    * spend `spender` tokens.
+    *
+    * Modifiers:
+    *  - onlySelf
+    */
+   function selfTransferFrom(
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol:326]: interface is still wrong

**Review by d10r — APPROVED** [review]
I renamed _holder_ to _sender_ in the interface, because that's how it's named elsewhere in the interface. Unfortunately we don't have consistent naming btw. interface and implementation file (see e.g. transferFrom(), thus there's no obvious best way. Should be addressed in the future, but not a blocker here.

**Comment by hellwolf** [comment]
> I renamed _holder_ to _sender_ in the interface, because that's how it's named elsewhere in the interface. Unfortunately we don't have consistent naming btw. interface and implementation file (see e.g. transferFrom(), thus there's no obvious best way. Should be addressed in the future, but not a blocker here.

indeed, naming is different in different contexts for good reason (transfers vs. approvals), but also different in different standards for bad reason (erc20 vs. erc777).

Changing interface name is fine, it shouldn't cause compatibility issues in web3 api, changing event though will affect web3 and also the subgraph.

Also note that the code is almost verbatim from openzeppellin erc777 of some point.

**Pull request merged**